src/azure-cli-core/azure/cli/core/auth/adal_authentication.py

```diff
@@ -18,9 +19,24 @@ class MSIAuthenticationWrapper(MSIAuthentication):
     def get_token(self, *scopes, **kwargs):  # pylint:disable=unused-argument
         logger.debug("MSIAuthenticationWrapper.get_token invoked by Track 2 SDK with scopes=%s", scopes)
 
+        # In Cloud Shell, we use
+        # - msrestazure to get access token
+        # - MSAL to get VM SSH certificate
         if 'data' in kwargs:
-            from azure.cli.core.azclierror import AuthenticationError
-            raise AuthenticationError("VM SSH currently doesn't support managed identity or Cloud Shell.")
+            if in_cloud_console():
+                import msal
+                from .util import check_result, build_sdk_access_token
+                app = msal.PublicClientApplication(
+                    "04b07795-8ddb-461a-bbee-02f9e1bf7b46",  # Use a real client_id, so that cache would work
+                    #token_cache=...,  # TODO: This PoC does not currently maintain a token cache;
+                                       #       Ideally we should reuse the real MSAL app object which has cache configured.
+                    )
+                result = app.acquire_token_interactive(list(scopes), prompt="none", data=kwargs["data"])
```

Comment:
@jiasli

* This draft has been updated to use the latest [MSAL-in-CloudShell calling pattern](https://github.com/AzureAD/microsoft-authentication-library-for-python/pull/420/files#diff-ed3b8af48cd21dc1ac735b839d89b172a6f45569c324af70a2bdb4f60e5fbcddR263-R271) and ready to be reviewed again.
* I manually tested this PoC by running "az ssh vm ..." inside my Cloud Shell environment which has the aforementioned latest MSAL. You could also create such a Cloud Shell environment for your testing, or you could wait several days until I release MSAL Python 1.18.
* Once MSAL Python 1.18 is released, you and/or I will modify the `setup.py` and `requirements.py3.Linux.txt` in this PR accordingly, and then change this PR from draft to a ready-to-merge PR, before 5/17.